### ISSUE TITLE
programmatic access to county-level covid data

### ISSUE THREAD

**Issue by paul-shannon** [mariam16548/learningShiny#6]
@mariam16548 First priority this week, I suggest, is to learn to call

  https://www.knowi.com/coronavirus-dashboards/covid-19-api/

from R.  

A useful learning task and demo could be this:  outside of shiny, create a bar chart of weekly infection for King County, using a live feed from knowi.com.  

You should  use the R package httr, which is introduced here:
      https://cran.r-project.org/web/packages/httr/vignettes/quickstart.html

PLEASE check in with me often!  This is new terrain for you, in several ways, and we want you to learn it quickly.  Which means:  read carefully, create simple examples (REALLY simple to start), notice when you are stuck, ask for help.

Speaking of simple examples, and building on your study of http a few weeks ago:

````
> library(httr)
> url <- "http://google.com"
> GET(url)
````

**Comment by mariam16548** [comment]
@paul-shannon No problem, I'll get to work on this. I'll let you know if I get stuck!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
So I've been reading through the link you sent me and learning more about httr, and so this is what I put into R: 
`library(httr)`
`url <- "https://knowi.com/api/data/ipE4xJhLBkn8H8jisFisAdHKvepFR5I4bGzRySZ2aaXlJgie?entityName=County%207%20day%20growth%20rates&exportFormat=csv"`
`GET(url)`

For some reason, my laptop is taking a really long time to export it into a .csv. Here is a screenshot of what it looks like on my end. 

![image](https://user-images.githubusercontent.com/52047042/89830911-d7d1cf80-db11-11ea-869f-113d47aa9b9d.png)

Any suggestions would be appreciated.

**Comment by paul-shannon** [comment]
a clue:  how big is the result?

**Comment by mariam16548** [comment]
@paul-shannon 
Okay, that makes sense-- I will wait and continue to learn more!

**Comment by paul-shannon** [comment]
and how specific is your query?

**Comment by paul-shannon** [comment]
call me?

**Comment by mariam16548** [comment]
Should I do this instead? 
`https://knowi.com/api/data/ipE4xJhLBkn8H8jisFisAdHKvepFR5I4bGzRySZ2aaXlJgie?entityName=County%207%20day%20growth%20rates&exportFormat=csv&c9SqlFilter=select%20*%20where%20County%20=%20King%20County`

**Comment by paul-shannon** [comment]
Take a look at this when you need to take a break.  It predicts deaths for the coming week (or is it just 4 days?) based on an ensemble of models, at different confidence intervals

  https://viz.covid19forecasthub.org/

Here is Washington state.    At 95% confidence, the range is pretty wide.

Two things of specific relevance to your project:

   1) prediction is made by combining multiple models  (you will eventually want to do the same)
   2) the prediction is presented as a confidence interval: 0%, 50%, 95%  (allows navigation of uncertainty)

 - Paul

**Comment by mariam16548** [comment]
Hi @paul-shannon , 
I just put up the code for my plot for the weekly infection for King County (kingCountyInfectionCount.R). I also put up my bar chart (kingCountyInfectionPlot.png). 
I'll continue looking at the link you sent me-- it looks really cool!

**Comment by paul-shannon** [comment]
Hi @mariam16548,   can't find it!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I just put it under the covidApp repo-- originally it was in the covidProject repo, which is maybe why you didn't see it? 
Here is the full path: covidProject/covidApp/kingCountyInfectionCount.R
Please let me know if you still can't find it.

**Comment by paul-shannon** [comment]
@mariam16548 
The missing file was my fault.  Thanks.
Your code does not run:

![image](https://user-images.githubusercontent.com/2480712/89950494-8f341800-dbde-11ea-88b1-d112d7d46faf.png)

**Comment by mariam16548** [comment]
Sorry about that, @paul-shannon. I just updated the code (there were just some small typos I missed).

**Comment by paul-shannon** [comment]
Good practice:  always pessimistically test code before it leaves your hands.

In this case you probably had data2 and data3 defined in your R session.

The pessimistic approach would have been to exit R, restart R, then run the code.

**Comment by mariam16548** [comment]
Good tip! I made sure to do that before I sent you this version of the code-- and I'll make sure to remember to do this in the future.

**Comment by paul-shannon** [comment]
Good work. 
Some suggestions:
Hadley Wickham's tidyverse is very popular, but it can be problematic.  It introduces new operators ````%>%```` and often gratuitously different functions for what base R does just fine.    And his "tibble" - it drives me crazy.  Base R's data.frame is good enough

I don't want to dissuade you from using Hadley's libraries  but I DO want you to be able to do things in base R with equal facility.

With that in mind, please rewrite your small script without library(tidyverse).

Please also use consistent formatting and operators.  For assignment, always use ````<-```` - avoid ````=````.
````data```` is an unhelpful variable name.  This is a tibble, make it instead a data.frame, and then call it something descriptive, like ````tbl.cases```` or ````tbl.kingCounty````

Then format your barplot function call like this, with spaces after the commas, and a newline,  for easier reading
````
barplot(tbl.counts$7_day_count",  main="Weekly COVID-19 Case Count in King County",
        xlab="Weeks since 1/29/20",  ylab="Weekly Infection Count")
````
**Every bit of code that you send me should be beautiful, consistent, and it should work.  Otherwise I have to spend time which more properly should have been spend by you!.  No messing around.  Make it highest quality every time**

Next up, write a function which, given a zip code, returns these data.frames.  You will need to find an API or data source which maps zip code to county.    Use unit tests.

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I just updated the code to reflect the changes you suggested: 
- I rewrote the script without the library(tidyverse)
- I renamed the "data" variables 
- I replaced all assignments with <- - 
- Reformatted the barplot function for easier reading
I'll keep working on the function that you requested and making sure to include unit tests!

**Comment by paul-shannon** [comment]
Thank you, Mariam!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I just added in my function countyCovidCasesPlot.R in the this path: covidProject/covidApp/countyCovidCasesPlot.R
Please let me know what you think!

**Comment by paul-shannon** [comment]
@mariam16548 - good work!   Good to see a unit test in the file.
Could you call when you get a chance?  I want to ask some leading questions - the goal of which is to give you practice decomposing code into lightweight functions, a step or two beyond what you have accomplished here.

**Comment by mariam16548** [comment]
@paul-shannon Okay, I'll call you around 3.

**Comment by mariam16548** [comment]
Hi @paul-shannon,
I just made a new repo (countyLevelInfectionData) with the new lightweight functions. I made separate files for the test_countyLevelInfectionData(), countyLevelInfectionData(), and zipcodeToCounty() since they are separate functions. 

You can find the files in this path: 
covidProject/countyLevelInfectionData/

I hope you have a great weekend!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I'm having some trouble with the integration testing. I've updated the shiny code so you can see what I have so far (covidProject/covidApp/covidSimpleCalculation.R). 

- I originally was using the `stopifnot()` function, but I wanted to have a "TRUE" output using `checkEquals()`.  
- One of my lines is: ` checkEquals(output$coloredBox$color, "orange")`. I get a strange error message that just says `Error:` (I also tried `browser()`, but that didn't give me any new information). 
Any suggestions are appreciated!

**Comment by paul-shannon** [comment]
Hi @mariam16548 

You already know how to do this!   You have a longish program that does not work.  What is the name of the strategy (the meta-strategy) I ALWAYS recommend?    How would you apply this here?

 - Paul

P.S.  I smelled trouble as soon as I saw that you were working in a "covid" directory.  Learning new, entirely novel coding techniques in the midst of an existing application is a recipe for heartache.

**Comment by mariam16548** [comment]
@paul-shannon 
Thank you for the reminder the mantra! I will continue to work on establishing steady ground and then slowing working up from there.

**Comment by paul-shannon** [comment]
yes - and divide (the problem into small separate bits) and conquer (each small bit which has only one mystery in it)

**Comment by mariam16548** [comment]
Hi @paul-shannon,
I hope you're doing well! 
I think I figured out where I went wrong. I need to add a couple more things to the code to make sure it is concise and easy to read. I should be done by tomorrow.

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
The code should be up now. I made sure to put it in a separate file so that we can make changes if necessary (covidProject/covidSimpleCalculationWithTest.R) without getting mixed up with the original working code without the test. Please let me know what you think!

**Comment by paul-shannon** [comment]
Hi Mariam,

Looks pretty good!

Two questions:

  1) how does it work?  is there a virtual web browser?  Please research this in detail.
  2) weren’t you going to use RUnit?

 - Paul

P.S.  I am away for the next 3 hours - will be glad to correspond or talk after I am back

**Comment by mariam16548** [comment]
@paul-shannon 
Okay, sounds good!

**Comment by paul-shannon** [comment]
For comparison, take a look at Selenium - a heavyweight solution commonly used with big commercial (non-shiny) webapps.

**Comment by mariam16548** [comment]
@paul-shannon 
I'll look into that!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
So here's where I'm at so far: 
- I edited the code to use RUnit-- in the examples I saw, they used `stopifnot()` so I thought that might have been standard practice for integration testing. I realized that using RUnit actually gives more detail in terms of the error messages, so I think I should just stick with RUnit.
- I also added some comments to the code to make sure I can show my understanding of the testing I used. 
- I'll keep learning about integration testing and Selenium!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
So I've been looking more into integration testing and with regards to the questions you asked: 
- How does it work?  Is there a virtual web browser? 
  - The idea behind integration testing is where individual units are combined and tested to make sure they are working properly when they integrated. 
  - It’s not working through a virtual web browser, it is completely server-side. This integration testing calculates the variables, given that some inputs have changed. No data is exchanged to the outside. 
  - I'm still learning more about Selenium, but from what I've read, it's a driver for web-app UI. I'm looking through RSelenium tutorials to try and understand more of how it can be used in R. 

I hope you have a wonderful weekend!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
So I finished reading through the book, and here are some of the ideas that stuck out to me: 
- Make sure a paragraph is used as a “unit of composition”. It should have a topic and concluding sentences to make it seem well-rounded and different from other paragraphs. 
- Use active voice since it is more direct and concise
- Put statements in positive form
  - “He was not very often on time” → “He usually came late” 
- Omit needless words 
- Vary sentence structure and length
- Parallel structure can sometimes be helpful 
  - “A time not for words, but action” → “A time not for words, but for action”
- Keep related words together in sentences
- He only found two mistakes vs. He found only two mistakes.
- Put the most prominent word/words of the sentence towards the end

I should be done writing the draft by tomorrow!

**Comment by mariam16548** [comment]
Hi @paul-shannon, 

I hope you had a great weekend! Over the weekend, I wanted to work a bit on the code we have so far: 

- For the `countyLevelInfectionData()` function, I added an error message for invalid zip codes. 
- I also added `startDate` and `endDate` arguments-- I made these arguments optional in that if the user doesn't put in any dates, it will default to showing all of the data.  (I also added error messages for dates that are unavailable.) I made tests to make sure that these arguments worked as expected. 
-  For the actual app, I wanted to incorporate the `countyLevelInfectionData()` function by adding in a numeric input for zip codes. I then used that data to produce COVID-19 cases based on the user's zip code. 
- Here's a screenshot of what I have so far. I do think it is a bit redundant to ask the users for the number cases/100,000 and then show them a graph of the number of cases. I'm thinking we can just use the data we have to answer the question for them. 
- You can find this code here: covidProject/covidAppV2/covidSimpleCalculationVersion2.R
![image](https://user-images.githubusercontent.com/52047042/92512708-0bad1d00-f1c4-11ea-8d8a-ed3461425e4b.png)

**Comment by paul-shannon** [comment]
Hi Mariam,

This looks good!

I agree: no need to ask users for information if you can obtain it from zip code (-> county -> recent data).

Let me know when you want me to publish the app.   This requires “dockerization” and will be available here:

    http://apps.systemsbiology.net/<someNameYouChoose>

 - Paul

P.S.  How goes the proposal revision (inspired, I hope, by the example I sent)?

**Comment by mariam16548** [comment]
@paul-shannon,
Thanks for the feedback! I'm still working on the proposal-- discussing some of the newer additions I've made to the app so far. I want to make it more thorough and follow the example you gave me. I'm also thinking about including a screenshot of what the app looks like so far so that the readers have an idea of what it looks like.

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I hope you're doing well. 

- I changed the app to only ask for zip code and population density (no longer includes a question about case count, as I was able to find it through the data we have).  
  -  covidProject/covidAppV2/covidSimpleCalculationVersion2.R
- I also have[ the proposal ](https://docs.google.com/document/d/1AF3zzW_1Pl0IfEfRNml9gJtxQ6rxzczWPWIwFlXHzmc/edit?usp=sharing)for you to take a look at when you have time.

**Comment by paul-shannon** [comment]
Can you deduce, infer population density from zip code?   Seem like there must be some programmatically queryable resource which provides that.

**Comment by mariam16548** [comment]
@paul-shannon 
I was going to work on that next! I'll let you know if I get stuck.

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I was able to infer the county population density from zip code-- finding the data was pretty challenging, but I found [this data](https://covid19.census.gov/datasets/21843f238cbb46b08615fc53e19e0daf/data?geometry=-126.247%2C28.795%2C126.878%2C67.148&selectedAttribute=B01001_calc_PopDensity) from the Census. 

- I made a `countyPopulationDensity` function that takes in the arguments of `countyName` and `stateName`, and it outputs the county population density using the data I found
- I edited the `countyLevelInfectionData` function to include the `countyPopulationDensity` function. 
  -  It outputs a table of the dates (first column), COVID-19 case counts (second column), and population density on the third column (new addition). 
- I edited the Shiny app to remove the question about population density and made sure to extract the information from the third column of the `countyLevelInfectionData()` output.
  -   You can see these Shiny app changes in this path: `covidProject/covidAppV2/covidSimpleCalculationVersion2.R`

**Comment by paul-shannon** [comment]
@mariam16548 good progress!
suggestions:
   - it doesn't make much sense to increment and decrement zip codes
   - better yet would be text entry widget, which only allows numbers, and which checks for validity when enough characters have been entered
  - remember that other factors go into your calculation:  masking, alcohol, group size, age (with loosely predicts risky behavior)

If you live in a high density, high infection rate county, but wear a mask and keep social distance, then your excursion out into the world is quite safe.   Please add these factors, incorporate them in the model, revise the app!

**Comment by mariam16548** [comment]
@paul-shannon 
Sounds great! I'll work on that!

**Comment by mariam16548** [comment]
@paul-shannon 
Progress so far: 
- I was able to change the zip code input to a text input instead of a numeric input. 
- I was able to add in the factors that you mentioned-- I'm still finding sources to try and come up with a reasonable calculation for how each of these factors affects the risk. 
- I was able to make my app work a bit faster by editing the functions I used to load the data before executing the function
- I hope to finalize all of these changes to the app and have them up by Monday.

Have a wonderful weekend!

**Comment by paul-shannon** [comment]
Good work, Mariam!

**Comment by paul-shannon** [comment]
Hi Mariam,

I have added comments to the first section of your proposal.  I urge you to refine the first two sections, distill them into a crystal clear and precise description of the project!   Boil it down to its essence.  Omit superfluous words and phrases.  

In my own writing, I often get lost in the sentences and then lose the paragraph - that is, I lose sight of the message to be conveyed. I lose the forest for the trees.

One technique for avoiding that is to create a minimalist statement, which can be for just one section.  For example, just the introduction.  Ask yourself: what is the bare-bones, most concise statement of my purpose in this paragraph?  Imagine you have 10 seconds to tell a stranger what you are working on.

For instance:
  * with only a vague sense of the risk, people can make bad covid-related choices
  * we quantify the risk by combining local data with their social plans

Work and rework the text till it is crystal clear!

 - Paul

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I just finished fixing the code. I added all of the changes I noted in the last comment as well as: 
- I fixed the integration test for the app to include multiple factors that I added (and the new risk calculation).
- I commented the sources for my risk calculation at the end of the app code 
  - I made sure that even if you live in a high density, high infection rate county, but wear a mask and keep social distance, then you should have a low-risk result. 
  - The app path is here: covidProject/covidAppV2/covidSimpleCalculationVersion2.R

I've looked over the comments on the proposal, and I'll get to work on them now!

**Comment by paul-shannon** [comment]
some quick comments:
   mask yes/no should be a radiobutton widget
   at least for now, please default the zip code to somewhere in seattle
   do you have any evidence for you risk scaling factors?  did you just make them up? :}

```separation of concerns```:  your risk calculation should not be in the GUI part of your program.  It is a separate and non-visual calculation.  It should leave in its own file, be extensible (add or replace components of your algorithm), have a dedicated set of unit tests with sample data.   These risk calculations are much studied, much modeled, and the subject of LOTS of research since February.   Please read up on them, write (probably just a skimpy prototype, proof-of-concept at first) a set of functions which draw upon at least one peer-reviewed and/or published study about transmission risk.

**Comment by mariam16548** [comment]
@paul-shannon 
I have a question: By "risk scaling factors", do you mean how I calculated groupSize risk factor?

**Comment by paul-shannon** [comment]
Hi Mariam,

Thank you for your patience with all my suggestions!

I mean all of your scaling operations, see below.
Several questions & suggestions here:

1) Always unwise to hard code constants into programming code.  Read them from a configuration file.
2) More fundamentally: where are these numbers from?  Based on what studies?  Tested how?  
3) Good engineering would have you define a RiskCalculator class, which would be assigned SpecificRiskXXXX objects.  These would not be part of your shiny app.   They would standalone, run alone, test alone, AND be included very simply via a function call in your app.

 - Paul

      
      if (masking=="Yes") {                    #decreased risk because more precautions are being taken
        likelihoodOfHarm<- likelihoodOfHarm*.35
      }  else {
        likelihoodOfHarm <- likelihoodOfHarm
      }
      
      if (age<30){
        likelihoodOfHarm <- likelihoodOfHarm*1.31     #increased risk because more likely to exhibit risky behavior
      }  else {
        likelihoodOfHarm <- likelihoodOfHarm
      }
      
      if (groupSize>10){
        likelihoodOfHarm <- likelihoodOfHarm*1.3 #increased risk because people around in close proximity
      } else if (groupSize>25){
        likelihoodOfHarm <- likelihoodOfHarm*1.6 #increased risk because more people around in close proximity
      } else if (groupSize>50){
        likelihoodOfHarm <- likelihoodOfHarm*1.9 #increased risk because more people around in close proximity
      }  else {
        likelihoodOfHarm <- likelihoodOfHarm
      }
      
      if (alcoholConsumption=="Yes"){
        likelihoodOfHarm <- likelihoodOfHarm*1.66 #increased risk because lack of awareness
      }  else {
        likelihoodOfHarm <- likelihoodOfHarm
      }

**Comment by paul-shannon** [comment]
It’s completely legit to “stub in” your risk calculation algorithms - which would eventually become nuanced and sophisticated over time, probably depending on multiple interacting factors.   Dead simple versions are fine to start.   

But the calculation should be separate code, in one or more separate files (one separate file is good to start). With tests.

I hope this makes sense!

 - Paul

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
Thanks for the clarification! I actually did base my risk scaling factors on specific sources I found (I put these at the bottom of the code, but I understand that we can make a separate configuration file where I can comment in these sources instead)
Here are a few of them: 
- [Source](https://www.thelancet.com/journals/lancet/article/PIIS0140-6736(20)31142-9/fulltext) for mask lowering risk calculation. I originally saw the statistic in [this](https://www.ucdavis.edu/coronavirus/news/your-mask-cuts-own-risk-65-percent/) article. 
- [Source](https://one.nhtsa.gov/people/injury/research/pub/HS809028/index_.htm) for alcohol causing lowered awareness and increasing risk calculation --> .08 BAC results in avg of 66% impairment
- Sources for younger people risk calculation:  
  -  [45%](https://www.usatoday.com/story/news/politics/2020/06/26/coronavirus-young-americans-less-likely-social-distance/3257513001/) of people under 30 are socializing with people they do not live with and not social distancing *(I still have to trace back where this news article got the data from)
  -   [around 40%](https://covid.idmod.org/data/Inequities_in_COVID-19_dynamics.pdf) of COVID cases are from people under 30. I originally saw the statistic [here](https://komonews.com/news/coronavirus/troubling-trend-shows-more-young-people-testing-positive-for-covid-19). 
  -   Therefore, around 16% of young people (likely asymptomatic) are engaging in risky behavior (.40*.45)
  - There is a [82% risk reduction](https://www.thelancet.com/journals/lancet/article/PIIS0140-6736(20)31183-1/fulltext) when social distancing 
  -  1-(.82*(1-.16))=.31 -->  31% risk increase for young people under the age of 30

**Comment by paul-shannon** [comment]
Good.  Thanks for the explanation.

Do you understand the need to put these in a separate, well-documented, tested file?  A software module?

**Comment by mariam16548** [comment]
@paul-shannon-- I do understand the need to put them in a separate file
I will get to work on making a function for the risk calculation (with tests) and including all of the sources as well.

**Comment by mariam16548** [comment]
Hi @paul-shannon, 
I just finished pushing the following changes to GitHub:  
- I wrote the `riskCalculation()` function (made the calculation shorter (not as redundant as before) and I added my sources there)
  -  I added test_riskCalculation.R as well
  -  I also updated the way I calculated the `groupSize` risk. I used the same method as this [event risk calculator](https://covid19risk.biosci.gatech.edu/). I found it to be a more accurate way of calculating how `groupSize` affects the risk (it required using county population and county case count). I already have case count data, but not county population (only population density).  
      -    So I made a new simple function `countyPopulation()` which uses Census data to find county population (input is `countyName` and `stateName`)      
   -   Path for risk calculation: covidProject/covidAppV2/riskCalculation.R

- For the Shiny app code, I used the radio buttons for yes/no questions and made the default location in Seattle
